tests/phpunit/tests/unicode/wpIsValidUtf8.php

```diff
@@ -0,0 +1,97 @@
+<?php
+/**
+ * Unit tests covering WordPress’ UTF-8 handling.
+ *
+ * @package WordPress
+ * @group unicode
+ */
+
+class Tests_WpIsValidUtf8TestCase extends WP_UnitTestCase {
+	/**
+	 * Verifies that WordPress can properly detect valid and invalid UTF-8.
+	 *
+	 * Ticket 38044
+	 *
+	 * @dataProvider data_utf8_test_data
+	 *
+	 * @param string $bytes Bytes as a PHP string.
+	 */
+	public function test_properly_validates_utf8( $bytes ) {
+		$is_valid = mb_check_encoding( $bytes, 'UTF-8' );
+
+		$this->assertSame(
+			$is_valid,
+			wp_is_valid_utf8( $bytes ),
+			$is_valid
+				? 'Should have identified the input as a valid UTF-8 string.'
+				: 'Should have rejected the input as a valid UTF-8 string.'
+		);
+	}
+
+	/**
+	 * Verifies that WordPress can properly detect valid and invalid UTF-8;
+	 * forces testing with the fallback mechanism in pure PHP code.
+	 *
+	 * Ticket 38044
+	 *
+	 * @dataProvider data_utf8_test_data
+	 *
+	 * @param string $bytes Bytes as a PHP string.
+	 */
+	public function test_fallback_properly_validates_utf8( $bytes ) {
+		$is_valid = mb_check_encoding( $bytes, 'UTF-8' );
+
+		$this->assertSame(
+			$is_valid,
+			_wp_is_valid_utf8_fallback( $bytes ),
+			$is_valid
+				? 'Should have identified the input as a valid UTF-8 string.'
+				: 'Should have rejected the input as a valid UTF-8 string.'
```

Comment:
Minor clarity improvement, "rejected… as a valid" sounds off:

```suggestion
				? 'Should have identified the input as a valid UTF-8 string.'
				: 'Should have rejected the invalid UTF-8 input string.'
```

Reply:
updated strings in e0fb4dab5d6d9115dc2172645df32b6775449a32